Code/GraphMol/SubstructLibrary/SubstructLibrary.h

```diff
@@ -0,0 +1,310 @@
+/*!
+  This is an API that hides the implementation details used for
+  indexing molecules for substructure searching.  It simply
+  provides an API for adding and getting molecules from a set.
+ */
+class MolHolderBase {
+public:
+  virtual ~MolHolderBase() {}
+
+  //! Add a new molecule to the substructure search library
+  //!  Returns the molecules index in the library
+  virtual unsigned int addMol( const ROMol &m ) = 0;
+  
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int) const = 0;
+     // throws IndexError?
+
+  //! Get the current library size
+  virtual unsigned int size() const = 0;
+};
+
+//! Concrete class that holds molecules in memory
+/*!
+    This is currently one of the faster implementations.
+    However it is very memory intensive.
+*/
+class MolHolder: public MolHolderBase {
+  std::vector<boost::shared_ptr<ROMol> > mols;
+public:
+  MolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+    mols.push_back( boost::make_shared<ROMol>(m) );
+    return size()-1;
+  }
+    
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    return mols[idx];
+  }
+  
+  virtual unsigned int size() const { return rdcast<unsigned int>(mols.size()); }
+
+};
+
+//! Concrete class that holds binary cached molecules in memory
+/*!
+  This implementation uses quite a bit less memory than the
+  non cached implementation.  However, due to the reduced speed
+  it should be used in conjunction with a pattern fingerprinter.
+
+  See RDKit::FPHolder
+*/
+class CachedMolHolder: public MolHolderBase {
+  std::vector<std::string> mols;
+public:
+  CachedMolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+    mols.push_back( std::string() );
+    MolPickler::pickleMol(m, mols.back());
+    return size()-1;
+  }
+
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    boost::shared_ptr<ROMol> mol(new ROMol);
+    MolPickler::molFromPickle(mols[idx], mol.get());
+    return mol;
+  }
+  
+  virtual unsigned int size() const { return rdcast<unsigned int>(mols.size()); }
+};
+
+//! Concrete class that holds smiles strings in memory
+/*!
+    This implementation uses quite a bit less memory than the
+    cached binary or uncached implementation.  However, due to the
+    reduced speed it should be used in conjunction with a pattern
+    fingerprinter.
+
+    See RDKit::FPHolder
+*/
+class CachedSmilesMolHolder: public MolHolderBase {
+  std::vector<std::string> mols;
+public:
+  CachedSmilesMolHolder() : MolHolderBase(), mols() {}
+
+  virtual unsigned int addMol( const ROMol &m ) {
+
+    bool doIsomericSmiles = true;
+    mols.push_back(MolToSmiles(m, doIsomericSmiles));
+    return size()-1;
+  }
+  
+  virtual boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
+    // throws IndexError?
+    boost::shared_ptr<ROMol> mol(SmilesToMol(mols[idx]));
```

Comment:
You had mentioned that before, are there examples of that?  Might be worth a note somewhere.  If this is the case, I expect it may be faster than the binary serialization.